### PR TITLE
add .abort() to the createAsyncThunk thunkAction

### PR DESCRIPTION
**Pull request by phryneas**
It's a bit... async, sorry - I don't think this is much simpler to solve.

The tests should give you a pretty good idea of what's actually happening.

### CODE REVIEW

**Comment by markerikson** [comment]
Yipes. Yeah, getting complicated.

It's almost like there are other asynchronous logic abstractions that are perhaps better meant for cancellation or something!

File: src/createAsyncThunk.ts [hunk 1]
```diff
@@ -129,21 +183,25 @@ export function createAsyncThunk<
...
-    return returned.payload
+    return Object.assign(thunkAction, { abort })
   }
 
   return Object.assign(actionCreator, {
     pending,
     rejected,
-    fulfilled,
-    unwrapResult
+    fulfilled
```
Comment (markerikson) [inline src/createAsyncThunk.ts]: Pretty sure we need to include `aborted` in here so it can be handled?

File: src/createAsyncThunk.ts [hunk 2]
```diff
@@ -84,24 +83,67 @@ export function createAsyncThunk<
     }
   )
 
+  const aborted = createAction(
```
Comment (markerikson) [inline src/createAsyncThunk.ts]: Do we actually need a separate `aborted` action type?

Could we use `rejected`, and have `aborted: true` as part of `meta`?

Comment (phryneas) [reply src/createAsyncThunk.ts]: That would be an option, too. I'm fine either way - your call :)

**Comment by phryneas** [comment]
> It's almost like there are other asynchronous logic abstractions that are perhaps better meant for cancellation or something!

Well, definitely not promises :D

This is what the fetch API uses, so it would fit most use cases I guess - and actually using it is pretty straightforward, too (see the tests).

File: src/createAsyncThunk.test.ts [hunk 3]
```diff
@@ -104,3 +109,111 @@ describe('createAsyncThunk', () => {
...
+      expect.any(Object),
+      expect.objectContaining({ type: 'test/pending' }),
+      expect.objectContaining({ type: 'test/fulfilled' })
+    ])
+  })
+
+  test('abort after dispatch', async () => {
+    const thunkAction = asyncThunk({})
+    const promise = store.dispatch(thunkAction)
+    thunkAction.abort('AbortReason')
```
Comment (markerikson) [inline src/createAsyncThunk.test.ts]: The usage here seems unlikely to occur in a real app.  Most code isn't going to reference the `thunkAction` value at all, as it's going to be `dispatch(asyncThunk())`.  

I realize that the abort handling needs to be on a per-call basis, so I'm not sure what the right answer is here.

File: etc/redux-toolkit.api.md [hunk 4]
```diff
@@ -102,32 +102,27 @@ export function createAction<P = void, T extends string = string>(type: T): Payl
 export function createAction<PA extends PrepareAction<any>, T extends string = string>(type: T, prepareAction: PA): PayloadActionCreator<ReturnType<PA>['payload'], T, PA>;
 
 // @alpha (undocumented)
-export function createAsyncThunk<ActionType extends string, Returned, ActionParams = void, TA extends AsyncThunksArgs<any, any, any> = AsyncThunksArgs<unknown, unknown, Dispatch>>(type: ActionType, payloadCreator: (args: ActionParams, thunkArgs: TA) => Promise<Returned> | Returned): ((args: ActionParams) => (dispatch: TA["dispatch"], getState: TA["getState"], extra: TA["extra"]) => Promise<import("./createAction").PayloadAction<Returned, string, {
+export function createAsyncThunk<ActionType extends string, Returned, ActionParams = void, TA extends AsyncThunksArgs<any, any, any> = AsyncThunksArgs<unknown, unknown, Dispatch>>(type: ActionType, payloadCreator: (args: ActionParams, thunkArgs: TA) => Promise<Returned> | Returned): ((args: ActionParams) => ((dispatch: TA["dispatch"], getState: TA["getState"], extra: TA["extra"]) => Promise<import("./createAction").PayloadAction<Returned, string, {
```
Comment (markerikson) [inline etc/redux-toolkit.api.md]: Missed the API file update again :)

File: src/createAsyncThunk.test.ts (hunk 3 above)
Comment (phryneas) [reply src/createAsyncThunk.test.ts]: Yeah, I think cancellation is not the most common thing per se, as for many people ignoring earlier running requests might be enough - but if needed this api provides all tools to create a simple reusable abstraction:

```js
function useAbortingApi(asyncApiAction, args) {
  const abortRef = useRef();
  const dispatch = useDispatch();
  return () => {
    const thunkAction = asyncApiAction(args);
    if (abortRef.current) abortRef.current();
    abortRef.current = thunkAction.abort;
    dispatch(thunkAction);
  }
}

function MyComponent(){
  const clickHandler = useAbortingApi(api.fetch, {});
  return <button onClick={clickHandler}>...</button>;
}
```

Comment (phryneas) [reply src/createAsyncThunk.test.ts]: That said, I *could* try to put an `abort` method on the promise returned by the thunk... Might even be pretty simple.

Comment (phryneas) [reply src/createAsyncThunk.test.ts]: That eliminated the "aborted before dispatch" case and allowed for a lot of simplifications.
I hadn't even considered that :man_facepalming:

Comment (phryneas) [reply src/createAsyncThunk.test.ts]: But the diff is ruined now *sigh*

File: src/createAsyncThunk.ts [hunk 5]
```diff
@@ -87,63 +86,94 @@ export function createAsyncThunk<
   const rejected = createAction(
     type + '/rejected',
     (error: Error, requestId: string, args: ActionParams) => {
+      if (error.name === 'AbortError') {
```
Comment (markerikson) [inline src/createAsyncThunk.ts]: Any chance we can collapse these two cases to shorten it a bit?

Comment (phryneas) [reply src/createAsyncThunk.ts]: Better?

**Comment by markerikson** [comment]
Awright, good enough for now.  Great work!

**Pull request merged**